### DISCUSSION TITLE
X/Twitter OAuth 2.0 User Context from Supabase

### DISCUSSION THREAD

**Discussion by degencodes** [supabase/supabase#29236] [Questions]
Is it possible to get X/Twitter OAuth 2.0 user context information from Supabase after logging in a user via Twitter? Looking to take Twitter actions on behalf of users. 

I can successfully sign in a user via Twitter in my web app, but when passing their provider_token as the Bearer in a Twitter api request, I get the error message :
"Authenticating with OAuth 2.0 Application-Only is forbidden for this endpoint.  Supported authentication types are [OAuth 1.0a User Context, OAuth 2.0 User Context]"

I saw [this post](https://github.com/orgs/supabase/discussions/2961 ) from 2021, but the most recent update suggests the issue hasn't been addressed.

**Comment by degencodes** [comment]
Bumping. Anyone know a bit about this?

**Reply by marcusmotill** [reply]
@degencodes looks like the supabase code for twitter is outdated and probably shouldn't be used. they are still using v1 and not oauth2. 

https://github.com/supabase/auth/blob/1c7202ff835856562ee66b33be131eca769acf1d/internal/api/provider/twitter.go#L18-L24
```ts
const (
	defaultTwitterAPIBase = "api.twitter.com"
	requestURL            = "/oauth/request_token"
	authenticateURL       = "/oauth/authenticate"
	tokenURL              = "/oauth/access_token" //#nosec G101 -- Not a secret value.
	endpointProfile       = "/1.1/account/verify_credentials.json"
)
```

whereas the [new examples](https://developer.x.com/en/docs/authentication/oauth-2-0/user-access-token) hit a `https://twitter.com/i/oauth2/*` endpoint that supabase hasn't chosen to add for some reason.

I think you'll need to roll your own auth flow for user facing X apis

**Reply by degencodes** [reply]
Thanks marcus. Ended up writing my own flow.

**Reply by runarlevi** [reply]
Hey! Im also looking into doing similar things. Was it hard to write your own flow?

**Reply by degencodes** [reply]
@runarlevi 
Not hard. The X api docs make it pretty straightforward. The biggest time sink was not realizing the new x auth wasn't compatible with supabase!

If anyone else stumbles across this discussion, good luck! Hopefully we get this upgrade in the future. I opened an issue for a feature request, but it was killed. https://github.com/supabase/supabase/issues/33641

There is a discussion post [here](https://github.com/orgs/supabase/discussions/2961) that has more traction.
